the-blue-alliance-ios/Protocols/Refreshable.swift

```diff
@@ -128,37 +128,44 @@ extension Refreshable {
 
     // TODO: Add a method to add an observer on a single core data object for changes
 
-    func cancelRefresh() {
-        if requests.isEmpty {
-            return
+    /**
+     * Add several operations to be executed in parallel. A `finish` block may be supplied to execute a block
+     * on the main thread after the passed operations have completed. If a `finish` block is supplied, the operation
+     * wrapping the `finish` block will be returned.
+     */
+    @discardableResult
+    func addRefreshOperations(_ operations: [Operation], finished: (() -> ())? = nil) -> Operation? {
+        // Create an operation to update our refresh indicator - should happen last.
+        let updateRefreshOperation = BlockOperation { [weak self] in
+            self?.updateRefresh()
+        }
+        var lastOperation = updateRefreshOperation
+
+        // If `finished` is supplied, create an operation to perform after all other operations are completed.
+        var finishedOperation: BlockOperation?
+        if let finished = finished {
+            let fo = BlockOperation(block: finished)
+            finishedOperation = fo
+            lastOperation = fo
+            updateRefreshOperation.addDependency(fo)
         }
 
-        for request in requests {
-            request.cancel()
+        for op in operations {
+            lastOperation.addDependency(op)
         }
-        requests.removeAll()
 
-        updateRefresh()
-    }
+        let mainThreadOperations = [updateRefreshOperation, finishedOperation].compactMap({ $0 })
+        OperationQueue.main.addOperations(mainThreadOperations, waitUntilFinished: false)
+        refreshOperationQueue.addOperations(operations, waitUntilFinished: false)
 
-    func addRequest(request: URLSessionDataTask) {
-        if requests.contains(request) {
-            return
-        }
-        requests.append(request)
         updateRefresh()
+
+        return finishedOperation
     }
 
-    func removeRequest(request: URLSessionDataTask) {
-        guard let index = requests.firstIndex(of: request) else {
-            return
-        }
-        requests.remove(at: index)
+    func cancelRefresh() {
+        refreshOperationQueue.cancelAllOperations()
         updateRefresh()
-
-        if requests.isEmpty {
-            noDataReload()
```

Comment:
Aaaa todo, we need to get this back in, and then remove where we do this manually.